src/mono/mono/mini/aot-compiler.c

```diff
@@ -6128,6 +6131,70 @@ method_is_externally_callable (MonoAotCompile *acfg, MonoMethod *method)
 	}
 }
 
+#ifdef MONO_ARCH_AOT_SUPPORTED
+static const char *
+get_pinvoke_import (MonoAotCompile *acfg, MonoMethod *method)
+{
+	MonoImage *image = m_class_get_image (method->klass);
+	MonoMethodPInvoke *piinfo = (MonoMethodPInvoke *) method;
+	MonoTableInfo *tables = image->tables;
+	MonoTableInfo *im = &tables [MONO_TABLE_IMPLMAP];
+	guint32 im_cols [MONO_IMPLMAP_SIZE];
+	char *import;
+
+	import = (char *)g_hash_table_lookup (acfg->method_to_pinvoke_import, method);
+	if (import != NULL)
+		return import;
+
+	if (piinfo->implmap_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_IMPLMAP, piinfo->implmap_idx))
+		return NULL;
+
+	mono_metadata_decode_row (im, piinfo->implmap_idx - 1, im_cols, MONO_IMPLMAP_SIZE);
+
+	int module_idx = im_cols [MONO_IMPLMAP_SCOPE];
+	if (module_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_MODULEREF, module_idx))
+		return NULL;
+
+	import = g_strdup_printf ("%s", mono_metadata_string_heap (image, im_cols [MONO_IMPLMAP_NAME]));
+
+	g_hash_table_insert (acfg->method_to_pinvoke_import, method, import);
+
+	return import;
+}
+#else
+static const char *
+get_pinvoke_import (MonoAotCompile *acfg, MonoMethod *method)
+{
+	return NULL;
+}
+#endif
+
+/*
+ * is_direct_pinvoke_specified_for_method:
+ *
+ * Returns whether the method is specified to be directly pinvoked based on
+ * the direct_pinvoke HashTable populated in process_specified_direct_pinvokes.
+ */
+static gboolean
+is_direct_pinvoke_specified_for_method (MonoAotCompile *acfg, MonoMethod *method)
+{
+	if (acfg->aot_opts.direct_pinvoke)
+		return TRUE;
+
+	if (!acfg->aot_opts.direct_pinvokes && !acfg->aot_opts.direct_pinvoke_lists)
+		return FALSE;
+
+	const char *sym = get_pinvoke_import (acfg, method);
+	const char *module_name = acfg->image->module_name;
```

Comment:
This is the module of the assembly that we currently AOT and not the module in the dllimport attribute. We probably need to extend get_pinvoke_import wit ability to extract both module as well as entrypoint, looks like implementation of get_pinvoke_import was originally from lookup_pinvoke_call_impl and there you see how it extracts the module part from the dllimport (used to load the shared library):

```
scope_token = mono_metadata_decode_row_col (mr, im_cols [MONO_IMPLMAP_SCOPE] - 1, MONO_MODULEREF_NAME);
orig_scope = mono_metadata_string_heap (image, scope_token);
```

You probably need to extend what's stored in the hash map, `acfg->method_to_pinvoke_import`, to hold both module and entrypoint and then change get_pinvoke_import to something like:

`gboolean get_pinvoke_import(MonoAotCompile *acfg, MonoMethod *method, const char **module, const char **entrypoint)`

and accept module and/or entrypoints to be NULL, meaning caller is not interested in getting that data back since we have callers only interested in entrypoint.

Also note that the moduleref_name is what's in the dllimport, so could be just module name, module name + file extension, full/absolute path, but for the direct pinvoke you would need to only strip it down to the module name without file extension.

Reply:
Modified `get_pinvoke_import` to `method_to_pinvoke_has_scope_import`, and as suggested, modified `acfg->method_to_pinvoke_import` to be a hashtable of form `MonoMethod:string array`, where the string array always has the scope first, then the import.

I'll still need to modify other calls to `method_to_pinvoke_has_scope_import` to properly free the memory allocated to hold the module and entrypoint strings. It seems like `get_pinvoke_import` didn't free the allocated string.

There some progress in the end to end testing at https://github.com/mdh1418/runtime/tree/mixed_reality_direct_pinvoke_end_to_end, I think I was able to find the symbol in the `iOS.Device.Aot-Llvm.Test.dll.s` file
```
adrp x16, mono_aot_iOS_Device_Aot_Llvm_Test_got@PAGE+0
add x16, x16, mono_aot_iOS_Device_Aot_Llvm_Test_got@PAGEOFF
ldr x0, [x16, #80]
.word 0xd63f0000
.word 0xaa0003fa
.word 0x910083a0
.word 0xf9400341
.word 0xf90013a1
.word 0xf9000340
.word 0x910063a0
bl _mono_threads_enter_gc_safe_region_unbalanced
.word 0xf90053a0
bl _hwang
.word 0xf94053a0
.word 0x910043a1
bl _mono_threads_exit_gc_safe_region_unbalanced
```

I still need to make sure to strip down the module name to be without the file extension

Reply:
The assembler you see there is the managed->native wrapper used for p/invokes and you see the direct call to your symbols `hwang` that has been prefixed since the platform uses a prefix for symbols, so that looks like expected.

Reply:
> Not needed, the ownership of the strings are kept in the hash table for the lifetime of the compile and that's why the parameter's are const char ** indicating to the caller that the returned strings are shared and should not be freed. That is how the current implementation of get_pinvoke_import works and that is how the new extended version should work as well.

Oh that makes sense, I had set the returned strings to be duplicates of the string values and missed that I could take advantage that the memory would be freed by the hashtable.

Reply:
Stripped down the module_ref to be the basename of the file, excluding any extensions